Fetching health data from Healthkit (Health app and Apple Watch)

mmervecerit opened the pull request:
# Fetching health data from Healthkit (Health app and Apple Watch)

## :recycle: Current situation & Problem
The goal was to get Heart rate, oxygen saturation, and body temperature from the HealthKit and saving it automatically  to Firebase.

## :gear: Release Notes 
*Add a bullet point list summary of the feature and possible migration guides if this is a breaking change so this section can be added to the release notes.*
*Include code snippets that provide examples of the feature implemented or links to the documentation if it appends or changes the public interface.*

Now we can get Heart rate, oxygen saturation, and body temperature in addition to wrist temperature (if provided through Apple Watch, cannot add manually)
`CollectSample(
                HKQuantityType(.heartRate),
                deliverySetting: .anchorQuery(.automatic)
            )
            CollectSample(
                HKQuantityType(.oxygenSaturation),
                deliverySetting: .anchorQuery(.automatic)
            )
            CollectSample(
                HKQuantityType(.appleSleepingWristTemperature),
                deliverySetting: .anchorQuery(.automatic)
            )
            CollectSample(
                HKQuantityType(.bodyTemperature),
                deliverySetting: .anchorQuery(.automatic))
        }`

## :white_check_mark: Testing
*Please ensure that the PR meets the testing requirements set by CodeCov and that new functionality is appropriately tested.*
*This section describes important information about the tests and why some elements might not be testable.*
Added health data manually to test the functionality. The wrist temperature cannot be tested since it doesn't allow to add values manually.

## :pencil: Code of Conduct & Contributing Guidelines 

By submitting creating this pull request, you agree to follow our [Code of Conduct](https://github.com/CS342/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/CS342/.github/blob/main/CONTRIBUTING.md):
- [x] I agree to follow the [Code of Conduct](https://github.com/CS342/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/CS342/.github/blob/main/CONTRIBUTING.md).

mmervecerit commented:
@nriedman, I can't understand why I am getting a build/Fastlane error here. Could this be somehow related to a previous TestFlight deployment error? Would love to get your support.

nriedman commented:
Looking into the failure, I'm noticing that the application is failing to launch for UI testing. We see the following error message:

```
The request to open "edu.stanford.cs342.2025.neutrofeverguard" failed.
Domain: FBSOpenApplicationServiceErrorDomain
Code: 4
User Info: {
    BSErrorCodeDescription = InvalidRequest;
    FBSOpenApplicationRequestID = 0xb080;
}

The operation couldn’t be completed. Application info provider (FBSApplicationLibrary) returned nil for "edu.stanford.cs342.2025.neutrofeverguard"
Domain: FBSOpenApplicationErrorDomain
Code: 4
Failure Reason: Application info provider (FBSApplicationLibrary) returned nil for "edu.stanford.cs342.2025.neutrofeverguard"
User Info: {
    BSErrorCodeDescription = NotFound;
}
```

Previously, we thought this might be an issue with our runner, but I'm starting to suspect it has to do with `Signing and Capabilities`. I tried manually adding the Bundle Identifier to the Info.plist and enabling automatic signing on the release target, but neither seems to fix the issue. Now, I'm wondering if the capabilities enabled here need a provisioning account -- a background fetch mode is enabled, and this might be more strictly controlled.

@PSchmiedmayer would love to hear your thoughts on this. #5 seems to have the same signing and capabilities but is passing the build and test. I'll continue looking into it.

nriedman commented:
Note: It looks like it's passing now for some reason, but the app is still failing to launch during the UI tests with the same error.

PSchmiedmayer commented:
@mmervecerit @nriedman We have taken a closer look at these issues and fixed them in our template application with the following PR: https://github.com/StanfordSpezi/SpeziTemplateApplication/pull/98.

The best way to replicate the changes would be to:
1. Update all dependencies to the latest version in Xcode and committing all the updated project files + Package.resolved file. You need to run **File** -> **Swift Packages** -> **Update to Latest Package Versions** in Xcode.
2. Modifying the UI tests to be a bit more stable and adjust to the latest changes in all the dependencies, mostly focusing on the diff in:
    a) [TemplateApplicationUITests/OnboardingTests.swift](https://github.com/StanfordSpezi/SpeziTemplateApplication/pull/98/files#diff-5be4df504b7d39125c283bea1051e3518f4cf9f60aacb32f3cd2a4d16b73edfc)
    b) [TemplateApplicationUITests/SchedulerTests.swift](https://github.com/StanfordSpezi/SpeziTemplateApplication/pull/98/files#diff-70ff708a6a827c4298bcdbcd74cc9b72cc6a5f2c8813c044c3087e8176938554)

The failing of launching the app is caused by the `app.deleteAndLaunch(withSpringboardAppName: "TemplateApplication")` which uninstalls the app and then re-installs the app. The logs about not being able to launch the app are triggered because the app has been deleted in the previous step and therefore Xcode re-installs it. We can discard these logs as they are by design and mean that we successfully uninstalled the app.

dusixian commented:
Hi Merve, thank you for your work on this! I have a few questions and suggestions:  

1. For fetching data from HealthKit, do you think a 1-month range is sufficient? The Health app allows viewing data by year, so should we consider fetching a longer history or even all available data?  
2. I noticed that you removed several empty lines in the following files:  
    - *NeutroFeverGuard/Firestore/FirebaseConfiguration.swift*  
    - *NeutroFeverGuard/HomeView.swift*  
    - *NeutroFeverGuard/NeutroFeverGuardDelegate.swift*  
   Was this change intended to pass any github test, or for code style consistency?
3. The scheme file (*NeutroFeverGuard.xcodeproj/xcshareddata/xcschemes/NeutroFeverGuard.xcscheme*) was modified in your commit. Would it be better to add this file to `.gitignore` so everyone can maintain their own scheme without affecting others?

mmervecerit commented:
@dusixian, thanks for your review. 

1. I asked Oliver about this, and he thought 1 month is a good timeline, so I kept it that way. We can change it pretty easily, but let's ask the project mentors for future iterations as well.
2. Yeah, this was mainly to get rid of some test issues. I did the same for the questionnaire as well, fixed it to pass some tests.
3. Sounds good; I added it to .gitignore, as suggested.

Thanks!

dusixian commented:
> @dusixian, thanks for your review.
> 
> 1. I asked Oliver about this, and he thought 1 month is a good timeline, so I kept it that way. We can change it pretty easily, but let's ask the project mentors for future iterations as well.
> 2. Yeah, this was mainly to get rid of some test issues. I did the same for the questionnaire as well, fixed it to pass some tests.
> 3. Sounds good; I added it to .gitignore, as suggested.
> 
> Thanks!

Thanks Merve! I would approve this pr.

dusixian reviewed (approved):
Asked about:
1. For fetching data from HealthKit, do you think a 1-month range is sufficient? The Health app allows viewing data by year, so should we consider fetching a longer history or even all available data?  
2. I noticed that you removed several empty lines in the following files:  
    - *NeutroFeverGuard/Firestore/FirebaseConfiguration.swift*  
    - *NeutroFeverGuard/HomeView.swift*  
    - *NeutroFeverGuard/NeutroFeverGuardDelegate.swift*  
   Was this change intentional for code style consistency, or was it an automatic formatting adjustment? Just checking to ensure alignment with our project's style guidelines.  
3. The scheme file (*NeutroFeverGuard.xcodeproj/xcshareddata/xcschemes/NeutroFeverGuard.xcscheme*) was modified in your commit. Would it be better to add this file to `.gitignore` so everyone can maintain their own scheme without affecting others?  

And got reply from Merve:
1. I asked Oliver about this, and he thought 1 month is a good timeline, so I kept it that way. We can change it pretty easily, but let's ask the project mentors for future iterations as well.
2. Yeah, this was mainly to get rid of some test issues. I did the same for the questionnaire as well, fixed it to pass some tests.
3. Sounds good; I added it to .gitignore, as suggested.

All recommended changes are made / explained. Approved.
